conda/core/solve.py

```diff
@@ -1029,6 +1039,250 @@ def _prepare(self, prepared_specs):
+    - Pruning the repodata (?) - JRG: Not sure if this happens internally or at all.
+    - Prioritizing different aspects of the solver (version, build strings, track_features...)
+    """
+
+    def __init__(self, prefix, channels, subdirs=(), specs_to_add=(), specs_to_remove=(),
+                 repodata_fn=REPODATA_FN, command=NULL):
+        super().__init__(prefix, channels, subdirs=subdirs, specs_to_add=specs_to_add,
+                         specs_to_remove=specs_to_remove, repodata_fn=repodata_fn,
+                         command=command)
+        self._state = {}
+
+    def solve_for_transaction(self, update_modifier=NULL, deps_modifier=NULL, prune=NULL,
+                              ignore_pinned=NULL, force_remove=NULL, force_reinstall=NULL,
+                              should_retry_solve=False):
+        # Logic heavily based on Mamba's implementation:
+        # https://github.com/mamba-org/mamba/blob/master/mamba/mamba.py#L289 (solver parts)
+
+        print("------ USING EXPERIMENTAL MAMBA INTEGRATIONS ------")
+
+        try:
+            # 1. Populate repos with installed packages
+            self._setup_state()
+            # 2. Create solver and needed flags, tasks and jobs
+            self._configure_solver()
+            # 3. Run the SAT solver
+            self._run_solver()
+            # 5. Export back to conda
+            return self._build_transaction()
+        finally:
+            # 6. Clean
+            self._state.clear()
+
+    def _setup_state(self):
+        from mamba.utils import load_channels, get_installed_jsonfile, init_api_context
+        from mamba.mamba_api import Pool, Repo, PrefixData
+
+        init_api_context()
+
+        pool = Pool()
+
+        # TODO: Check if this update-related logic is needed here too
+        # Maybe conda already handles that beforehand
+        # https://github.com/mamba-org/mamba/blob/master/mamba/mamba.py#L426-L485
+
+        prefix_data = PrefixData(self.prefix)
+        prefix_data.load()
+        installed_json_f, installed_pkgs = get_installed_jsonfile(self.prefix)
+
+        repos = []
+        installed = Repo(pool, "installed", installed_json_f.name, "")
+        installed.set_installed()
+        repos.append(installed)
+
+        # This function will populate the pool/repos with
+        # the current state of the given channels
+        channels = [c.canonical_name for c in self.channels]
+        index = load_channels(pool, channels, repos)
+
+        self._state.update({
+            "pool": pool,
+            "prefix_data": prefix_data,
+            "repos": repos,
+            "index": index,
+            "installed_pkgs": installed_pkgs,
+        })
+
+        return self._state
+
+    def _configure_solver(self):
+        if self.specs_to_add or context.update_modifier == UpdateModifier.UPDATE_ALL:
+            return self._configure_solver_for_install()
+        elif self.specs_to_remove:
+            return self._configure_solver_for_remove()
```

Comment:
Can we ever have both? The legacy solver seemed to be able to handle that from looking at the API, but I might be wrong.
Still, it would probably be a good idea to error out when we get the two actions, as we do not actually support them. Currently, if resolving for install, `self.specs_to_remove` will be silently ignored, which I imagine could lead to nightmare debugging :sweat_smile:

Reply:
Not sure about this. Mamba does not allow both operations, I think. The API supports that because some solves might involve installing/updating stuff _and_ removing existing packages. However, the user will only be able to say "add this" or "remove this", but never "remove this and add this" at the same time. 

Anaconda should confirm anyway because I am not sure.

And yes, please check for that silent ignore, I forgot about it assuming this is coming straight from the CLI, but who knows how the Python API will be used.